crates/proc_macro_api/src/version.rs

```diff
@@ -0,0 +1,131 @@
+    let mut items = ver_str.split_whitespace();
+    let tag = items.next().ok_or(err!("version format error"))?;
+    if tag != "rustc" {
+        return Err(err!("version format error (No rustc tag)"));
+    }
+
+    let version_part = items.next().ok_or(err!("no version string"))?;
+    let mut version_parts = version_part.split("-");
+    let version = version_parts.next().ok_or(err!("no version"))?;
+    let channel = version_parts.next().unwrap_or_default().to_string();
+
+    let commit = items.next().ok_or(err!("no commit info"))?;
+    // remove (
+    if commit.len() == 0 {
+        return Err(err!("commit format error"));
+    }
+    let commit = commit[1..].to_string();
+    let date = items.next().ok_or(err!("no date info"))?;
+    // remove )
+    if date.len() == 0 {
+        return Err(err!("date format error"));
+    }
+    let date = date[0..date.len() - 2].to_string();
+
+    let version_numbers = version
+        .split(".")
+        .map(|it| it.parse::<usize>())
+        .collect::<Result<Vec<_>, _>>()
+        .map_err(|_| err!("version number error"))?;
+
+    if version_numbers.len() != 3 {
+        return Err(err!("version number format error"));
+    }
+    let version = (version_numbers[0], version_numbers[1], version_numbers[2]);
+
+    Ok(RustCInfo { version, channel, commit, date })
+}
+
+// This is used inside read_version() to locate the ".rustc" section
+// from a proc macro crate's binary file.
+fn read_section<'a>(dylib_binary: &'a [u8], section_name: &str) -> io::Result<&'a [u8]> {
+    BinaryFile::parse(dylib_binary)
+        .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?
+        .section_by_name(section_name)
+        .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidData, "section read error"))?
+        .data()
+        .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))
+}
+
+// Check the version of rustc that was used to compile a proc macro crate's
+// binary file.
+// A proc macro crate binary's ".rustc" section has following byte layout:
+// * [b'r',b'u',b's',b't',0,0,0,5] is the first 8 bytes
+// * ff060000 734e6150 is followed, it's the snappy format magic bytes,
+//   means bytes from here(including this sequence) are compressed in
+//   snappy compression format. Version info is inside here, so decompress
+//   this.
+// The bytes you get after decompressing the snappy format portion has
+// following layout:
+// * [b'r',b'u',b's',b't',0,0,0,5] is the first 8 bytes(again)
+// * [crate root bytes] next 4 bytes is to store crate root position,
+//   according to rustc's source code comment
+// * [length byte] next 1 byte tells us how many bytes we should read next
+//   for the version string's utf8 bytes
+// * [version string bytes encoded in utf8] <- GET THIS BOI
+// * [some more bytes that we don really care but still there] :-)
+// Check this issue for more about the bytes layout:
+// https://github.com/rust-analyzer/rust-analyzer/issues/6174
```

Comment:
Can you use doc comments instead of regular comments where applicable. They show up even for private items when using `--document-private-items`.